[WIP] feat(wallet): CIP-36 Voting formalisation

tomislavhoracek opened the pull request:
# Context

TBD

# Proposed Solution

# Important Changes Introduced

mkazlauskas commented on packages/wallet/src/SingleAddressWallet.ts:
```diff
@@ -273,6 +287,38 @@ export class SingleAddressWallet implements ObservableWallet {
+      address: ownAddress,
+      value: { coins: 1_000_000n }
+    }]);
+    const { constraints, utxo, implicitCoin, validityInterval, changeAddress } = await this.#prepareTx({
```
Why not just use `this.initializeTx`? If we can do that then entire `initializeVotingRegistrationTx` is not very useful and it would be enough to export a util to `constructVotingAuxData`.

mkazlauskas commented on packages/wallet/src/SingleAddressWallet.ts:
```diff
@@ -383,4 +429,38 @@ export class SingleAddressWallet implements ObservableWallet {
   }
+
+  // TODO - maybe we can move this to keyAgent? + cleanup
+  async #constructVotingAuxData(props: VotingAuxDataProps): Promise<Cardano.AuxiliaryData> {
```
How about moving this and voting constants under `wallet/Transaction`? This function can take in a KeyAgent as a dependency argument.

mkazlauskas reviewed (changes requested):
Good start. Besides the comments there are some lint issues to resolve.

tomislavhoracek commented on packages/wallet/src/KeyManagement/types.ts:
```diff
@@ -156,3 +156,8 @@ export interface KeyAgent {
 } & Shutdown;
+
+export interface VotingKeyPair {
+  prvKey: CSL.Bip32PrivateKey;
```
Already changed :) Will be introduced with next commit

mkazlauskas commented on packages/wallet/src/KeyManagement/util/constants.ts:
```diff
@@ -11,3 +11,8 @@ export enum Cip1852PathLevelIndexes {
   INDEX = 4
 }
+
+export enum VotingLabels {
```
I'm not sure this belongs under `KeyManagement/util` - it's specific to constructing cip-compliant transaction, not key management in general. How about moving cip36-specific stuff under `src/cip36`?

mkazlauskas commented on packages/wallet/src/KeyManagement/types.ts:
```diff
@@ -151,8 +158,20 @@ export interface KeyAgent {
+  /**
+   * @throws AuthenticationError
+   */
+  signVotingMetadata(props: SignVotingMetadataProps): Promise<Cardano.Ed25519Signature>;
```
`KeyAgent` interface is currently modeled to have a static `purpose`, `keyType` and `account` which doesn't match the up-and-coming [voting cip proposal](https://github.com/input-output-hk/gov-delegation-tools/blob/main/src/voting/CIP-voting.md) (it expects to wallet to use a different `purpose` etc.).

I suggest removing this and:
- extend `SerializableInMemoryKeyAgentData` to support using different `purpose` and `keyType`
- add util to create voting-specific `SerializableInMemoryKeyAgentData` (takes in `encryptedRootPrivateKeyBytes` and `networkId`), so it can be derived from the same root private key if desired.
- add util to `signVotingMetadata(keyAgent: KeyAgent, props: SignVotingMetadataProps)` (separate from KeyAgent interface or any specific key agent implementation) that uses `KeyAgent.signBlob` to do the actual signing
- add optional `purpose`, `keyType` and `accountIndex` in `KeyAgent.signBlob` params - it could be used for validating that correct instance of InMemoryKeyAgent is being used to sign voting metadata

mkazlauskas commented on packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts:
```diff
@@ -49,9 +54,39 @@ export class InMemoryKeyAgent extends KeyAgentBase {
     this.#getPassword = getPassword;
   }
 
+  async squashAccount({ purpose, coinType, accountIndex }: AccountDerivationPath): Promise<KeyAgent> {
```
See update in my comment [here](https://github.com/input-output-hk/cardano-js-sdk/pull/278/files/5f9ed2439d8e07a991dc6f0fcca40b5dcf0d155c#r915778092). I suggest changing this to just `exportKeyPair(fullDerivationPath: number[])`

tomislavhoracek replied:
As we agreed, we will remove this method `squashAccount `

tomislavhoracek replied:
https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/eff6dece9e1d44c14975347f6b856983a716bad7

tomislavhoracek commented on packages/wallet/src/KeyManagement/types.ts:
```diff
@@ -125,12 +148,25 @@ export interface SignTransactionOptions {
 export interface KeyAgent {
   get networkId(): Cardano.NetworkId;
   get accountIndex(): number;
+  get purpose(): number;
+  get coinType(): number;
```
As we agreed, I will do that as the last thing if there will be time

tomislavhoracek commented on packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts:
```diff
@@ -49,9 +54,39 @@ export class InMemoryKeyAgent extends KeyAgentBase {
+      rootPrivateKey
+    });
+    const password = await getPasswordRethrowTypedError(this.#getPassword);
+    const encryptedRootPrivateKey = await emip3encrypt(rootPrivateKey.as_bytes(), password);
```
https://github.com/input-output-hk/cardano-js-sdk/pull/278#discussion_r915866817

tomislavhoracek commented on packages/wallet/src/GovernanceApi.ts:
```diff
@@ -0,0 +1,113 @@
+export class GovernanceApi {
+  readonly #keyAgent: AsyncKeyAgent;
+
+  constructor(keyAgent: AsyncKeyAgent) {
```
we will keep using generic GovernanceApi with no keyAgent limits

tomislavhoracek commented on packages/wallet/src/GovernanceApi.ts:
```diff
@@ -0,0 +1,113 @@
+    rewardAccountKey,
+    nonce
+  }: SignDelegationProps): Promise<Cardano.Ed25519Signature> {
+    const blake2b = require('blake2b');
```
Done in https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/eff6dece9e1d44c14975347f6b856983a716bad7

tomislavhoracek commented on packages/wallet/src/KeyManagement/types.ts:
```diff
@@ -125,12 +148,25 @@ export interface SignTransactionOptions {
+  /**
+   * @throws AuthenticationError
+   */
+  squashAccount(derivationPath: AccountDerivationPath): Promise<KeyAgent>;
```
Removed in https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/eff6dece9e1d44c14975347f6b856983a716bad7

tomislavhoracek commented:
@mkazlauskas 
I rebased to master and resolved conflicts. So, specific commit no. I added to CR comments also changed.
- All changes related to the latest review applied in [chore(wallet): changes related to the latest review](https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/0b66f2c5421fd389c5d6343cc339fdcc3663214f)
- Small improvement / fix where I added new keyAgent method to web-extension props [chore(web-extension): exportKeyPair added to keyAgentProperties](https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/8b23a5d4cb74407f1752fa7b05a365a65b42f20c)
- e2e fix - I fixed the voting test since it was moved to the e2e package after rebasing. It is tested and green! https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/67505bd2f0db8fec733c74ec33cb94b4c1d94a21

mkazlauskas commented on packages/wallet/src/KeyManagement/types.ts:
```diff
@@ -187,8 +188,15 @@ export interface KeyAgent {
+  /**
+   * @throws AuthenticationError
+   */
+  exportKeyPair(fullDerivationPath: number[]): Promise<KeyPair>;
```
Let's not add this to KeyAgent interface. Instead, just on InMemoryKeyAgent. Then we won't need the commit in web-extension package.

tomislavhoracek replied:
@mkazlauskas Isn't it better to keep it because if you use the wrong key agent e.g. LedgerKeyAgent we will throw a nice error that the operation is not supported by that key agent. We will also need those methods in other key agents because voting keys will be derived on a different way for HW ( we will see how ) or we will need to change approach once CIP proposal is approved?

tomislavhoracek commented on packages/wallet/src/KeyManagement/InMemoryKeyAgent.ts:
```diff
@@ -164,6 +140,24 @@ export class InMemoryKeyAgent extends KeyAgentBase {
+      .derive(harden(fullDerivationPath[Cip1852PathLevelIndexes.PURPOSE]))
+      .derive(harden(fullDerivationPath[Cip1852PathLevelIndexes.COIN_TYPE]))
+      .derive(harden(fullDerivationPath[Cip1852PathLevelIndexes.ACCOUNT]))
+      .derive(harden(fullDerivationPath[Cip1852PathLevelIndexes.ROLE]))
```
Just to be sure. So, we will just provide key/pair for `purpose/cointype/account` or you can pass FULL path and you will get key / pair also for that?

mkazlauskas replied:
My idea here is that `fullDerivationPath` can be any length

tomislavhoracek replied:
👌

tomislavhoracek replied:
Done in [chore(wallet): accept variable derivation path in exportKeyPair](https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/284de67cb69e01203a8dbeb0e4b4d9771f696b08)

tomislavhoracek commented on packages/wallet/src/GovernanceApiBase.ts:
```diff
@@ -14,37 +17,36 @@ export interface SignDelegationProps {
-      prvKey: Cardano.Bip32PrivateKey(cslPrvKeyBytes),
-      pubKey: Cardano.Bip32PublicKey(cslPubKeyBytes)
+      rawVotingPubKey,
+      skey,
```
@mkazlauskas I was also thinking about that but then I decided to expose also key / pair. Maybe end apps like Lace will need that. What you think?

mkazlauskas replied:
I think we should follow governance api here. Wallet will also have access to underlying `this.#getVotingKeyPair` if it needs it

tomislavhoracek replied:
👌

tomislavhoracek replied:
@mkazlauskas added in:
[chore(wallet): added toEd25519KeyPair util](https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/3b0cae214936fc0e75e3c762fc426f6c530aef8d) AND [5267630](https://github.com/input-output-hk/cardano-js-sdk/pull/278/commits/5267630966ba4899d57ad24ee0037822b7afaf16)

mkazlauskas commented on packages/web-extension/src/keyAgent/util.ts:
```diff
@@ -5,6 +5,8 @@ export const keyAgentChannel = (walletName: string) => `${walletName}$-keyAgent`
 export const keyAgentProperties: RemoteApiProperties<KeyManagement.AsyncKeyAgent> = {
   deriveAddress: RemoteApiPropertyType.MethodReturningPromise,
+  derivePublicKey: RemoteApiPropertyType.MethodReturningPromise,
+  exportKeyPair: RemoteApiPropertyType.MethodReturningPromise,
```
This is no longer a property of AsyncKeyAgent

mkazlauskas commented on packages/e2e/test/wallet/SingleAddressWallet/voting.test.ts:
```diff
@@ -0,0 +1,93 @@
+    const loadedTxAuxBlobSignature = loadedTx.auxiliaryData?.body.blob?.get(
+      BigInt(VotingLabels.SIG)
+    ) as Cardano.MetadatumMap;
+    const votingSignature = auxBlobData.get(1n) as Buffer;
```
`Buffer` is `Uint8Array`, but better use the narrower type here that is used in `core` [type](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/core/src/Cardano/types/AuxiliaryData.ts#L9).
```suggestion
    const votingSignature = auxBlobData.get(1n) as Uint8Array;
```

mkazlauskas commented on packages/wallet/src/GovernanceApiBase.ts:
```diff
@@ -0,0 +1,101 @@
+
+export type GetVotingKeyPair = (derivationPath: number[]) => Promise<KeyPair>;
+
+export class GovernanceApiBase implements GovernanceApi {
```
Tests are structured differently here - they're under `test/GovernanceApi`. Let's put this file and associated types under `src/GovernanceApi`

mkazlauskas replied:
```suggestion
export class GovernanceApi implements CipXXXXGovernanceApi {
```

mkazlauskas commented on packages/wallet/src/types.ts:
```diff
@@ -102,3 +103,9 @@ export interface ObservableWallet {
+export interface GovernanceApi {
+  getVotingKey(derivationPath: number[]): Promise<Ed25519KeyPair>;
+  buildDelegation(props: BuildDelegationProps): Promise<Cardano.AuxiliaryData>;
+  signDelegation(props: SignDelegationProps): Promise<Cardano.Ed25519Signature>;
```
Shouldn't this take in the same thing that `buildDelegation` returns?

```suggestion
  signDelegation(props: Cardano.AuxiliaryData): Promise<Cardano.Ed25519Signature>;
```

Then signDelegation implementation can just use `coreToCsl.txMetadatum` to serialize it

mkazlauskas commented on packages/wallet/src/types.ts:
```diff
@@ -102,3 +103,9 @@ export interface ObservableWallet {
   shutdown(): void;
 }
+
+export interface GovernanceApi {
```
```suggestion
export interface CipXXXXGovernanceApi {
```

mkazlauskas commented on packages/wallet/src/KeyManagement/types.ts:
```diff
@@ -49,16 +59,40 @@ export enum CommunicationType {
+export interface GroupedAddress {
   networkId: Cardano.NetworkId;
-  accountIndex: number;
+  derivationPath: AccountDerivationPath & {
```
This will be a breaking change with loading from pouchdb stores. Let's rename the store to `'addresses_v2'` [here](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/wallet/src/persistence/pouchdbStores/pouchdbWalletStores.ts#L50)

rhyslbw replied:
This is now a non-issue right?

rhyslbw reviewed (approved):
Nice refactor @mkazlauskas. Having a clear definition of the CIP is critical, and that's been achieved here. Please squash and let's get this merged!
